### PR TITLE
Switch to CBV2G from OpenV2G for DIN, ISO-2, PnC and SAP

### PR DESCRIPTION
**Pull request by SiebrenW**
This is the initial work to switch from OpenV2G to CBV2G in the EvseV2g module.

Currently The following is done:

- [x] add cbv2g as a dependency
- [x] link cbv2g to EvseV2G module
- [x] switch SAP
- [x] switch DIN
- [x] switch ISO -2
- [x] switch xmldsig fragment encoding
- [x] Remove OpenV2G
- [x] Test din
- [x] Test iso2
- [x] Test PnC

### CODE REVIEW

File: modules/EvseV2G/CMakeLists.txt
Comment (SebaLukas) [inline modules/EvseV2G/CMakeLists.txt:35]: The module does not require ISO20

File: .clangd
Comment (SebaLukas) [inline .clangd:1]: You can add this file to .gitignore but dont commit this kind of file

**Review by SebaLukas — COMMENTED** [review]
For now, the PR looks really good. I have just successfully started and finished a DIN session in the SIL. Now the code has to be tested a lot in the SIL and on real cars. I'll be doing that over the next few weeks

File: .clangd
Comment (SiebrenW) [reply .clangd:1]: Sorry, that was by accident. I'll just move it down.

File: modules/EvseV2G/iso_server.cpp
Comment (SebaLukas) [inline modules/EvseV2G/iso_server.cpp:2372]: -1 causes a Failed_UnknowSession, as the byte length is set to 7. And the car sends a false SessionID with only 7 bytes. Therefore delete -1.

Comment (SiebrenW) [reply modules/EvseV2G/iso_server.cpp]: Good catch.
Fixed. That creapt in because I wasn't sure about the null terminator initially. Removed it for the chars, but I must have added it to the bytes accidentally.

File: modules/EvseV2G/din_server.cpp
Comment (SiebrenW) [inline modules/EvseV2G/din_server.cpp:924]: Should remove

**Review by SebaLukas — COMMENTED** [review]
I found a few little things

**Comment by SebaLukas** [comment]
I was able to successfully test DIN and ISO-2 AC & DC EIM with the SIL. Only with PnC there seems to be a problem with the authorization. I'll have a look at that soon. Once this has been resolved, I would also test this PR on a few cars.

Please update/rebase your branch to eversest-core main. Then there should be no more conflict with the dependencies.yaml.

**Comment by SiebrenW** [comment]
I had a few errors building with main, which is why this was rebased to the release initially. I didn't think it was going to cause merge conflicts in the actual module code.
I will rebase to main hopefully later today (and cleanup those comments).

Then the iso2_ParameterSetType_5_ARRAY_SIZE being insufficient, I would prefer to keep using the constants from cbv2g, so I will wait for that update.
I marked master as the dependency tag, so we will see a compile error once this number (and thus the name) has changed and fix it then and there.

**Comment by SebaLukas** [comment]
You can also just clone everest-core and during cmake `edm` will take care of pulling the right dependencies with the right versions. Then building should not really be a problem with different dependency versions.

I totally agree with `iso2_ParameterSetType_5_ARRAY_SIZE`. This should be fixed in libcbv2g and not here. That was actually just a reminder for me not to forget this.

**Comment by SiebrenW** [comment]
I will try again then when I get home. Last time I tried to build master as is it didn't like my effort.
Or do you mean without everest-workspace? The readme does not cite this as an option, so I never tried.

**Comment by SebaLukas** [comment]
You have two options: 
Like the readme says, create a everest-workspace with edm init and get all necessary repos/libs in one folder.

Or you can clone only the repos you work on. For me it is usually:
```
.
├── everest-core
├── Josev
└── libiso15118
```
First, edm checks whether all necessary repos are stored locally in the same folder as everest-core. If not, then edm download the missing repos in the $CPM_SOURCE_CACHE folder or build folder.

The second option should be somewhere in our documentation.

File: modules/EvseV2G/v2g.hpp
Comment (SiebrenW) [reply modules/EvseV2G/v2g.hpp:258]: I see that I did use a variable that is not comparable directly now. I will fix that too, but won't rebase it for now.
Though before I don't see what the relation is between iso1ServiceListType_Service_ARRAY_SIZE and iso1ServiceParameterListType either.

The direct alternative to the OpenV2G iso1ServiceListType_Service_ARRAY_SIZE is iso2_ServiceType_8_ARRAY_SIZE.
Why? It has little to do with each other. A hardcoded 8 would make more sense, or am I missing some relation between the size of the array of an array of ServiceParameters versus an array of ParameterSets.

Comment (SiebrenW) [reply modules/EvseV2G/v2g.hpp]: I found the reason. in v2g_ctx.cpp add_service_to_service_list the same index is used for both the evse_service_list and service_parameter_list.
Weird choice, but okay.

**Comment by SiebrenW** [comment]
Question: does it make sense to enforce the requirement of unused variables in the SignedInfo type?
For instance to set ANY_isUsed = 0; in SignatureMethod?
I made use of the fact that the object needs no conversion, it's the same one, but the conversion did set some of these fields to 0 explicitly.
I will add this in another commit for now. We can always drop it.

**Comment by SebaLukas** [comment]
Yes, add that :+1:

**Comment by SebaLukas** [comment]
I activated the pr actions (A maintainer have to approve this for forks :shrug:)
The clang-format check failed. Can you run clang-format on your changes (in your everest-core folder) with:
```
docker run \
--volume "$(pwd):/source
ghcr.io/everest/everest-clang-format:latest \
-r /source
```
This Docker image is also used here in the actions. This ensures that the same clang-format version is used.

**Comment by SiebrenW** [comment]
> I activated the pr actions (A maintainer have to approve this for forks 🤷) The clang-format check failed. Can you run clang-format on your changes (in your everest-core folder) with:
> 
> ```
> docker run \
> --volume "$(pwd):/source
> ghcr.io/everest/everest-clang-format:latest \
> -r /source
> ```
> 
> This Docker image is also used here in the actions. This ensures that the same clang-format version is used.

The command has some issues. The quote isn't even closed and when I do add it docker can't find the image.

File: modules/EvseV2G/iso_server.cpp
Comment (SebaLukas) [inline modules/EvseV2G/iso_server.cpp:227]: Please change the 8 to a zero. That was the reason why PnC didn't work. The Exi fragment was encoded with an 8 byte offset, so of course the digest value was no longer correct.

**Comment by SebaLukas** [comment]
Now PnC works also in the SIL. All that's left now are the tests with real cars.

**Comment by SebaLukas** [comment]
> 
> The command has some issues. The quote isn't even closed and when I do add it docker can't find the image.
> 
> I did the clang check manually with git-clang-format HEAD~4. I fixed it in one of the commits.

Sorry about the incorrect clang-format cmd. I always use podman, so my cmd is slightly different.

That should work here:
```
docker run --user "1000:100" --volume "$(pwd):/source" http://ghcr.io/everest/everest-clang-format:latest -r /source --extensions "hpp,cpp" -e "/source/cache" -e "/source/build" -e "/source/build-cross" -e "/source/build-clang" -i
```

Docker should find the image, I was able to pull it normally with podman earlier.

File: modules/EvseV2G/iso_server.cpp
Comment (SiebrenW) [reply modules/EvseV2G/iso_server.cpp:227]: Well. That's embarressing.
I didn't even copy paste that. I actually typed 8 for some reason.

**Comment by SiebrenW** [comment]
The 8 has be turned into a 0 🤦 
If the SIL was better documented I would have tested this myself. I am already happy to be able to test DC BPT with nodered.

Docker is still complaining with docker: invalid reference format. I think my git-clang-tidy was effective anyways.

**Comment by barsnick** [comment]
> ~Then the iso2_ParameterSetType_5_ARRAY_SIZE being insufficient, I would prefer to keep using the constants from cbv2g, so I will wait for that update.~ (or not [#563 (comment)](https://github.com/EVerest/everest-core/pull/563#discussion_r1516819337) )
> I marked master as the dependency tag, so we will see a compile error once this number (and thus the name) has changed and fix it then and there.

https://github.com/EVerest/cbexigen/pull/74

**Comment by corneliusclaussen** [comment]
Will tag as post-release, I think we need a lot more testing and it should be merged after the march stable release

**Comment by SebaLukas** [comment]
Todo from my side before merge: 1. Fix integration test 2. squash all commits together

**Comment by SebaLukas** [comment]
Hi :)

unfortunately I haven't had time to test your PR with real cars yet.
In a week and a half the Charin Testival will take place in Poland and some of my Pionix colleagues will be there, too.

I would also like to test your PR with the cars on site.

I cloned your fork earlier and updated locally your branch to the current main. And created an extra module “EvseV2GLibCbV2G” with a copy of your EvseV2G module. I would then use this module for testing. Then it is easy to quickly change and test the modules through the config.

If there are no problems at the testival, then we can definitely merge your PR afterwards.

**Comment by SebaLukas** [comment]
We were actually able to test everything well at the Testival (except PnC) and we only found one error. I was able to fix this error straight away. We were able to successfully test PnC with an Ioniq6 last week. So nothing speaks against a merge :+1: 

I would update this branch in the next few days and add the fixes we needed for the Testival. And merge this PR after that.

I hope that I can create another release for the libcbv2g this week, then we can use the new libcbv2g version `0.1.1` right away.

**Comment by SebaLukas** [comment]
For me, nothing speaks against a merge. @corneliusclaussen & @barsnick could you take another look and see if you notice anything?

File: modules/EvseV2G/v2g_ctx.cpp
Comment (barsnick) [inline modules/EvseV2G/v2g_ctx.cpp:11]: Alternatively
```suggestion
#include <cbv2g/iso_2/iso2_msgDefDatatypes.h>
```

File: modules/EvseV2G/v2g_server.cpp
Comment (barsnick) [inline modules/EvseV2G/v2g_server.cpp:382]: Might use `exi_bitstream_reset()`, but okay if it works.

**Review by barsnick — COMMENTED** [review]
Thanks for the big effort!

It looks pretty good from what I can tell. I cannot judge on the fragment (signature) encoding, but fine since you tested it.

Only small remarks, mostly regarding the #include style.

File: modules/EvseV2G/v2g_server.cpp
Comment (SebaLukas) [reply modules/EvseV2G/v2g_server.cpp:382]: Good point :+1: But I think we can delete line 381 and 382 because in line 383 `exi_bitstream_init()` is called. In the init function `bit_count` and `byte_pos` is set to 0 anyway.

File: modules/EvseV2G/iso_server.cpp
Comment (SebaLukas) [reply modules/EvseV2G/iso_server.cpp:21]: Good point. I will change the three includes to `#include <`.

File: modules/EvseV2G/v2g_ctx.cpp
Comment (SebaLukas) [reply modules/EvseV2G/v2g_ctx.cpp]: Done

File: modules/EvseV2G/iso_server.cpp
Comment (SebaLukas) [reply modules/EvseV2G/iso_server.cpp]: Done

File: modules/EvseV2G/v2g_server.cpp
Comment (SebaLukas) [reply modules/EvseV2G/v2g_server.cpp]: Done

File: modules/EvseV2G/iso_server.cpp
Comment (barsnick) [inline modules/EvseV2G/iso_server.cpp]: Move this up with the other includes? I would have thought clang-format might automatically re-order anyway (in a different way, BTW) - therefore the PR check will fail anyway. ;-)

File: modules/EvseV2G/v2g_ctx.cpp
Comment (barsnick) [inline modules/EvseV2G/v2g_ctx.cpp]: Move this up with the other includes? I would have thought clang-format might automatically re-order anyway (in a different way, BTW) - therefore the PR check will fail anyway. ;-)

File: modules/EvseV2G/iso_server.cpp
Comment (SebaLukas) [reply modules/EvseV2G/iso_server.cpp]: I noticed that, too. This should be fixed with the recent commit

File: modules/EvseV2G/v2g_ctx.cpp
Comment (SebaLukas) [reply modules/EvseV2G/v2g_ctx.cpp]: I noticed that, too. This should be fixed with the recent commit

File: modules/EvseV2G/iso_server.cpp
Comment (barsnick) [inline modules/EvseV2G/iso_server.cpp:24]: Now I don't know what the h*ll clang-format is trying to do. :-D
#include "iso_server.hpp"
belongs further down, to where it used to be, above log.hpp.

**Review by barsnick — APPROVED** [review]
Just the minor #include nit.

File: modules/EvseV2G/iso_server.cpp
Comment (SebaLukas) [reply modules/EvseV2G/iso_server.cpp:24]: Yes, that was me, I'll change it back.

**Review by SebaLukas — APPROVED** [review]
:+1:

**Review by corneliusclaussen — APPROVED** [review]
Looks good to me! We also did quite some funtional testing here.

**Comment by SebaLukas** [comment]
I would now squash the commits together and then merge them. Thank you @SiebrenW  for creating this PR :smiley:

**Pull request merged**
